### PR TITLE
Handle a few edge cases for phpstan level 7

### PR DESCRIPTION
**Pull request by AJenbo**
Type: refactoring / documentation update
Breaking change: no

Handle a few minor issues. This solves about 20% of the remaining issues found by PHPStan at level 7.

I used array shapes in the PHPDoc to better document some of the function signatures.

Not sure why tests are failing, it appears unrelated to my changes here.

### CODE REVIEW

**Review by kylekatarnls — APPROVED** [review]

File: src/main/php/PDepend/Report/Dependencies/Xml.php [hunk 1]
```diff
@@ -306,6 +306,10 @@ protected function writeNodeDependencies(DOMElement $xml, AbstractASTArtifact $n
         }
 
         $xml = end($this->xmlStack);
+        if (!$xml) {
+            return;
```
Comment (kylekatarnls) [inline src/main/php/PDepend/Report/Dependencies/Xml.php:310]: `writeNodeDependencies` is called in only one place:

```php
        array_push($this->xmlStack, $typeXml);
        $this->writeNodeDependencies($typeXml, $type);
```
And `$typeXml` is `DOMElement` (because it's `DOMElement|false` on creation then `->setAttribute()` is called on it which means that if it's false, we have a PHP error earlier.

+ we get `$xml = end($this->xmlStack)` but it gives us exactly the same as `$xml` as already passed (which is unused as it's immediately re-set.

So this path is never taken unless someone extended this class and directly call it. But this method is so non-intuitive that I really doubt of it.

TL;DR It's a complete mess, I prefer to fix, rather than we do try to just tackle them with early return just for the sake of making PHPStan happy.

I propose we simply use `$xml` parameter and don't read the stack.

**Review by kylekatarnls — DISMISSED** [review]

**Review by kylekatarnls — APPROVED** [review]
